de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/session/internal/SharedReferencePointMapper.java

```diff
@@ -0,0 +1,543 @@
+ */
+class SharedReferencePointMapper {
+
+    private static final Logger LOG = Logger
+        .getLogger(SharedReferencePointMapper.class);
+
+    /**
+     * Mapping from reference point IDs to currently registered shared reference
+     * point.
+     */
+    private Map<String, IReferencePoint> idToReferencePointMapping;
+
+    /**
+     * Mapping from currently registered shared reference point to their id's.
+     */
+    private Map<IReferencePoint, String> referencePointToIDMapping;
+
+    /**
+     * Map for storing which clients have which reference point. Used by the
+     * host to determine who can currently process an activity related to a
+     * particular reference point. (Non-hosts don't maintain this map.)
+     */
+    private Map<User, List<String>> referencePointsOfUsers;
+
+    /**
+     * Map for storing the set of resources shared for each shared reference
+     * point. Maps to <code>null</code> for completely shared reference points.
+     */
+    private Map<IReferencePoint, Set<IResource>> partiallySharedResourceMapping;
+
+    /**
+     * Set containing the currently completely shared reference points.
+     */
+    private Set<IReferencePoint> completelySharedReferencePoints;
+
+    /**
+     * Set containing the currently partially shared reference points.
+     */
+    private Set<IReferencePoint> partiallySharedReferencePoints;
+
+    public SharedReferencePointMapper() {
+        idToReferencePointMapping = new HashMap<String, IReferencePoint>();
+        referencePointToIDMapping = new HashMap<IReferencePoint, String>();
+        referencePointsOfUsers = new HashMap<User, List<String>>();
+        partiallySharedResourceMapping = new HashMap<IReferencePoint, Set<IResource>>();
+        completelySharedReferencePoints = new HashSet<IReferencePoint>();
+        partiallySharedReferencePoints = new HashSet<IReferencePoint>();
+    }
+
+    /**
+     * Adds a reference point to the set of currently shared reference points.
+     * <p>
+     * It is possible to "upgrade" a partially shared reference point to a
+     * completely shared reference point by just adding the same reference point
+     * with the same ID again that must now marked as not partially shared.
+     * </p>
+     * 
+     * @param id
+     *            the ID for the reference points
+     * @param referencePoint
+     *            the reference points to add
+     * @param isPartially
+     *            <code>true</code> if the reference point should be treated as
+     *            a partially shared reference point, <code>false</code> if it
+     *            should be treated as completely shared
+     * 
+     * @throws NullPointerException
+     *             if the ID or reference point is <code>null</code>
+     * @throws IllegalStateException
+     *             if the ID is already in use or the reference point was
+     *             already added
+     */
+    public synchronized void addReferencePoint(String id,
```

Comment:
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Refactor this method to reduce its Cognitive Complexity from 19 to the 15 allowed. ([squid:S3776](https://sonarcloud.io/organizations/saros/rules#rule_key=squid%3AS3776))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=saros_complete&pullRequest=257&issues=AWbeHvN4SgN4uCclRsG1&open=AWbeHvN4SgN4uCclRsG1)

[ISSUE_KEY]: AWbeHvN4SgN4uCclRsG1